File: 3427-special-array-ii/3427-special-array-ii.cs
public class Solution {
    public bool[] IsArraySpecial(int[] nums, int[][] queries) {
        int n = nums.Length;
        int q = queries.Length;
        int[] hell = new int[n];
        for (var i = 1; i < n; ++i)
            hell[i] = (((nums[i] ^ nums[i - 1]) & 1) != 0) ? hell[i - 1] : i;
        
        bool[] ans = new bool[q];
        for (int i = 0; i < q; ++i) {
            int f =  queries[i][0];
            int t =  queries[i][1];
            ans[i] = hell[t] <= f;
        }
        return ans;
    }
}

File: 3427-special-array-ii/3427-special-array-ii.java
class Solution {
    public boolean[] isArraySpecial(int[] nums, int[][] q) {
        int l = nums.length;
        int[] arr = new int[l];
        for (int i = 1; i < l; ++i)
            arr[i] = (((nums[i] ^ nums[i - 1]) & 1) != 0) ? arr[i - 1] : i;
            
        int q_len = q.length;
        boolean[] ans = new boolean[q_len];
        for (int i = 0; i < q_len; ++i) {
            int f = q[i][0];
            int t = q[i][1];
            ans[i] = arr[t] <= f;
        }
        return ans;
    }
}